chainntnfs/txnotifier_test.go

```diff
@@ -604,6 +605,112 @@ func TestTxNotifierFutureSpendDispatch(t *testing.T) {
+// script multiple times).
+func TestTxNotifierFutureConfDispatchReuseSafe(t *testing.T) {
+	t.Parallel()
+
+	type status struct {
+		confirmed bool
+		done      bool
+	}
+	cancelCtx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+
+	// Create a listener function that just accepts and records event
+	// notifications in a loop that can be cancelled once the test is done.
+	notificationListener := func(ntfn *chainntnfs.ConfRegistration,
+		status *status) {
+
+		for {
+			select {
+			case <-ntfn.Event.Updates:
+			case <-ntfn.Event.Confirmed:
+				status.confirmed = true
+			case <-ntfn.Event.Done:
+				status.done = true
+			case <-cancelCtx.Done():
+				break
+			}
+		}
+	}
+
+	currentBlock := uint32(10)
+	hintCache := newMockHintCache()
+	n := chainntnfs.NewTxNotifier(
+		currentBlock, 2, hintCache, hintCache,
+	)
+
+	// We'll register a TX that sends to our test script and put it into a
+	// block.
+	tx1 := wire.MsgTx{Version: 1}
+	tx1.AddTxOut(&wire.TxOut{PkScript: testRawScript})
+	tx1Hash := tx1.TxHash()
+	ntfn1, err := n.RegisterConf(&tx1Hash, testRawScript, 1, 1)
+	if err != nil {
+		t.Fatalf("unable to register ntfn: %v", err)
+	}
+	status1 := &status{false, false}
+	go notificationListener(ntfn1, status1)
+
+	block := btcutil.NewBlock(&wire.MsgBlock{
+		Transactions: []*wire.MsgTx{&tx1},
+	})
+	currentBlock++
+	err = n.ConnectTip(block.Hash(), currentBlock, block.Transactions())
+	if err != nil {
+		t.Fatalf("unable to connect block: %v", err)
+	}
+	if err := n.NotifyHeight(currentBlock); err != nil {
+		t.Fatalf("unable to dispatch notifications: %v", err)
+	}
+
+	// Now register a second TX that spends to the same output and get it
+	// confirmed.
+	tx2 := wire.MsgTx{Version: 1}
+	tx2.AddTxOut(&wire.TxOut{PkScript: testRawScript})
+	tx2Hash := tx2.TxHash()
+	ntfn2, err := n.RegisterConf(&tx2Hash, testRawScript, 1, 1)
+	if err != nil {
+		t.Fatalf("unable to register ntfn: %v", err)
+	}
+	status2 := &status{false, false}
+	go notificationListener(ntfn2, status2)
+
+	block2 := btcutil.NewBlock(&wire.MsgBlock{
+		Transactions: []*wire.MsgTx{&tx2},
+	})
+	currentBlock++
+	err = n.ConnectTip(block2.Hash(), currentBlock, block2.Transactions())
+	if err != nil {
+		t.Fatalf("unable to connect block: %v", err)
+	}
+	if err := n.NotifyHeight(currentBlock); err != nil {
+		t.Fatalf("unable to dispatch notifications: %v", err)
+	}
+
+	// Finally, mine a few empty blocks and expect both TXs to be confirmed.
+	for currentBlock < 15 {
+		block := btcutil.NewBlock(&wire.MsgBlock{})
+		currentBlock++
+		err = n.ConnectTip(block.Hash(), currentBlock, block.Transactions())
+		if err != nil {
+			t.Fatalf("unable to connect block: %v", err)
+		}
+		if err := n.NotifyHeight(currentBlock); err != nil {
+			t.Fatalf("unable to dispatch notifications: %v", err)
+		}
+	}
+
+	// If we reach this point without panic, the bug this test was written
+	// for no longer exists.
+	if !status1.confirmed || !status2.confirmed {
```

Comment:
Yes, I just saw the race test failure today.
We need to read multiple events from the channels because there will be multiple confirmations the way it is implemented at the moment. So I solved it with wait groups for now.

But maybe that's another reason to go with the other implementation where we clean out all scripts that have been confirmed and therefore would only send one confirmation event.

Reply:
Yeah, I think that's the way it should be fixed.